Some multi-GPU unit tests hang when running in a different Docker environment

https://github.com/dmlc/xgboost/pull/5873#issuecomment-666696157
Log from 4-process setup:
```
pid = 21916, Device = 0
pid = 21919, Device = 0
pid = 21921, Device = 0
pid = 21924, Device = 0
```
~even though here we should have been using 4 GPUs.~ See https://github.com/dmlc/xgboost/issues/5963#issuecomment-666721876.

The undefined behavior exists in following tests:
* `tests/python-gpu/test_gpu_with_dask.py::TestDistributedGPU::test_dask_array`
* `tests/distributed/runtests-gpu.sh`

The behavior is "undefined" in the sense that using a different Docker container causes the tests to fail, even though they were succeeding previously.

Passing (current CI setup):
```bash
tests/ci_build/ci_build.sh gpu_build_centos6 docker --build-arg CUDA_VERSION=10.0 \
  tests/ci_build/build_via_cmake.sh  -DUSE_CUDA=ON -DUSE_NCCL=ON \
  -DOPEN_MP:BOOL=ON -DHIDE_CXX_SYMBOLS=ON -DGPU_COMPUTE_VER=75
tests/ci_build/ci_build.sh gpu nvidia-docker -it --build-arg CUDA_VERSION=10.2 \
  tests/ci_build/test_python.sh mgpu
```

Failing (the test just hangs):
```bash
tests/ci_build/ci_build.sh gpu_build docker --build-arg CUDA_VERSION=10.2 \
  tests/ci_build/build_via_cmake.sh  -DUSE_CUDA=ON -DUSE_NCCL=ON \
  -DOPEN_MP:BOOL=ON -DHIDE_CXX_SYMBOLS=ON -DGPU_COMPUTE_VER=75
tests/ci_build/ci_build.sh gpu nvidia-docker -it --build-arg CUDA_VERSION=10.2 \
  tests/ci_build/test_python.sh mgpu
```

I found the error by setting env var `DMLC_WORKER_STOP_PROCESS_ON_ERROR=false`:
```
/home/ubuntu/.local/lib/python3.7/site-packages/xgboost/core.py:1161: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise XGBoostError(py_str(_LIB.XGBGetLastError()))
E   xgboost.core.XGBoostError: [09:04:03] ../src/tree/updater_gpu_hist.cu:723: Exception in gpu_hist: NCCL failure :unhandled system error ../src/common/device_helpers.cu(71)
E   
E   Stack trace:
E     [bt] (0) /home/ubuntu/.local/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0xa4924) [0x7f40db349924]
E     [bt] (1) /home/ubuntu/.local/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x45a5a3) [0x7f40db6ff5a3]
E     [bt] (2) /home/ubuntu/.local/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x177dd2) [0x7f40db41cdd2]
E     [bt] (3) /home/ubuntu/.local/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x179e65) [0x7f40db41ee65]
E     [bt] (4) /home/ubuntu/.local/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x1a352b) [0x7f40db44852b]
E     [bt] (5) /home/ubuntu/.local/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(XGBoosterUpdateOneIter+0x69) [0x7f40db33dc39]
E     [bt] (6) /opt/python/envs/gpu_test/lib/python3.7/lib-dynload/../../libffi.so.6(ffi_call_unix64+0x4c) [0x7f43633f5630]
E     [bt] (7) /opt/python/envs/gpu_test/lib/python3.7/lib-dynload/../../libffi.so.6(ffi_call+0x22d) [0x7f43633f4fed]
E     [bt] (8) /opt/python/envs/gpu_test/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(_ctypes_callproc+0x2e7) [0x7f436340b6d7]

/home/ubuntu/.local/lib/python3.7/site-packages/xgboost/core.py:188: XGBoostError
```

Commands:
```bash
tests/ci_build/ci_build.sh gpu_build docker --build-arg CUDA_VERSION=10.2 \
  tests/ci_build/build_via_cmake.sh  -DUSE_CUDA=ON -DUSE_NCCL=ON \
  -DOPEN_MP:BOOL=ON -DHIDE_CXX_SYMBOLS=ON -DGPU_COMPUTE_VER=75
CI_DOCKER_EXTRA_PARAMS_INIT='-e DMLC_WORKER_STOP_PROCESS_ON_ERROR=false' \
  tests/ci_build/ci_build.sh gpu nvidia-docker -it --build-arg CUDA_VERSION=10.2  \
  tests/ci_build/test_python.sh mgpu
```

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -368,8 +368,10 @@ def TestPythonGPU(args) {
     def docker_args = "--build-arg CUDA_VERSION=${args.host_cuda_version}"
     if (args.multi_gpu) {
       echo "Using multiple GPUs"
+      // Allocate extra space in /dev/shm to enable NCCL
+      def docker_extra_params = "CI_DOCKER_EXTRA_PARAMS_INIT='--shm-size=4g'"
       sh """
-      ${dockerRun} ${container_type} ${docker_binary} ${docker_args} tests/ci_build/test_python.sh mgpu
+      ${docker_extra_params} ${dockerRun} ${container_type} ${docker_binary} ${docker_args} tests/ci_build/test_python.sh mgpu
       """
     } else {
       echo "Using a single GPU"

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -89,8 +89,8 @@ pipeline {
         script {
           parallel ([
             'test-python-cpu': { TestPythonCPU() },
-            'test-python-gpu-cuda10.0': { TestPythonGPU(host_cuda_version: '10.0') },
             'test-python-gpu-cuda10.2': { TestPythonGPU(host_cuda_version: '10.2') },
+            'test-python-gpu-cuda11.0-cross': { TestPythonGPU(artifact_cuda_version: '10.0', host_cuda_version: '11.0') },
             'test-python-gpu-cuda11.0': { TestPythonGPU(artifact_cuda_version: '11.0', host_cuda_version: '11.0') },
             'test-python-mgpu-cuda10.2': { TestPythonGPU(artifact_cuda_version: '10.2', host_cuda_version: '10.2', multi_gpu: true) },
             'test-cpp-gpu-cuda10.2': { TestCppGPU(artifact_cuda_version: '10.2', host_cuda_version: '10.2') },

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -92,7 +92,7 @@ pipeline {
             'test-python-gpu-cuda10.0': { TestPythonGPU(host_cuda_version: '10.0') },
             'test-python-gpu-cuda10.2': { TestPythonGPU(host_cuda_version: '10.2') },
             'test-python-gpu-cuda11.0': { TestPythonGPU(artifact_cuda_version: '11.0', host_cuda_version: '11.0') },
-            'test-python-mgpu-cuda10.2': { TestPythonGPU(host_cuda_version: '10.2', multi_gpu: true) },
+            'test-python-mgpu-cuda10.2': { TestPythonGPU(artifact_cuda_version: '10.2', host_cuda_version: '10.2', multi_gpu: true) },
             'test-cpp-gpu-cuda10.2': { TestCppGPU(artifact_cuda_version: '10.2', host_cuda_version: '10.2') },
             'test-cpp-gpu-cuda11.0': { TestCppGPU(artifact_cuda_version: '11.0', host_cuda_version: '11.0') },
             'test-jvm-jdk8-cuda10.0': { CrossTestJVMwithJDKGPU(artifact_cuda_version: '10.0', host_cuda_version: '10.0') },

